src/iter/collect/mod.rs

```diff
@@ -19,8 +18,10 @@ where
 {
     v.truncate(0); // clear any old data
     let mut collect = Collect::new(v, pi.len());
-    pi.drive(collect.as_consumer());
-    collect.complete();
+    let writes = pi.drive(collect.as_consumer());
```

Comment:
My only hesitance is that tracking `writes` correctly is very important for safety, but this PR moves that value through the control of a generic unknown third party, `pi`.

I'm not sure how bad that is. `pi.drive` has to be generic in the consumer type, which means it's generic in the `C::Result` type too, so it doesn't *know* that we're expecting a `usize`. Thus it can't directly misbehave by feeding us bad values. I think the worst *innocent* bug would be neglecting to reduce the results from two consumer splits, and then we'd just panic for not having enough writes.

A nefarious `pi` might be able to cheat in other ways, like `unsafe` code, or maybe even safe specialization on the `C::Result` type. But I doubt this could be done on accident, and we probably don't need to worry about whether someone actively shoots their own foot.

So... maybe it's fine, but I wanted to raise the point and give it some thought.

Reply:
So, this is admittedly a largely theoretical concern, but I think it's a real one, and I'm inclined not to make this change for this reason.

As you note, it's certainly possible for `pi` to have bugs and to return the wrong value to us, whether that be by failing to reduce properly or by using specialization. 

I agree it seems unlikely but those unlikely circumstances are exactly the ones that take the longest to track down when they happen. =) 

I could imagine trying to declare the traits to be unsafe, but it'd be a real shame. I remember the bad old days when rayon had unsafe all over the place: getting rid of that was always a nice benefit.

If there is a performance concern, I suspect we could create a "tree" of atomic integers and recoup similar benefits that way, without requiring as must trust.

Reply:
I'm confused as to what can be specialised to make bad things happen here. Is it about specialising with `usize` consumer results? If so, can't we just make `Collect` use a private wrapper around `usize`? You need to be able to name the things for which you want to specialise, right?